libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/map/MapEventDelegate.kt

```diff
@@ -0,0 +1,97 @@
+import com.mapbox.maps.plugin.gestures.OnMoveListener
+import com.mapbox.maps.plugin.gestures.addOnMapClickListener
+import com.mapbox.maps.plugin.gestures.addOnMapLongClickListener
+import com.mapbox.maps.plugin.gestures.addOnMoveListener
+import com.mapbox.maps.plugin.locationcomponent.OnIndicatorPositionChangedListener
+import com.mapbox.maps.plugin.locationcomponent.location
+import com.mapbox.navigation.utils.internal.InternalJobControlFactory
+import com.mapbox.navigation.utils.internal.ifNonNull
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.launch
+
+class MapEventDelegate(private val mapView: MapView) {
+
+    private val jobControl = InternalJobControlFactory.createMainScopeJobControl()
+
+    private val _mapStyleSink: MutableSharedFlow<Style> = MutableSharedFlow(
+        replay = 1,
+        onBufferOverflow = BufferOverflow.DROP_OLDEST,
+        extraBufferCapacity = 1
+    )
+    val mapStyleUpdates: Flow<Style> = _mapStyleSink
+
+    private val _mapLongClickSink: MutableSharedFlow<Point> = MutableSharedFlow(
+        extraBufferCapacity = 1,
+        onBufferOverflow = BufferOverflow.DROP_OLDEST
+    )
+    val mapLongClicks: Flow<Point> = _mapLongClickSink
+
+    private val _mapClickSink: MutableSharedFlow<Point> = MutableSharedFlow(
+        extraBufferCapacity = 1,
+        onBufferOverflow = BufferOverflow.DROP_OLDEST
+    )
+    val mapClicks: Flow<Point> = _mapClickSink
+
+    private val _positionChangeSink: MutableSharedFlow<Point> = MutableSharedFlow()
+    val positionChanges: Flow<Point> = _positionChangeSink
+
+    private val _mapMovementsSink: MutableSharedFlow<CameraState> = MutableSharedFlow(
+        extraBufferCapacity = 1,
+        onBufferOverflow = BufferOverflow.DROP_OLDEST,
+    )
+    val mapMovements: Flow<CameraState> = _mapMovementsSink
+
+    private val onStyleLoadedListener = OnStyleLoadedListener {
+        ifNonNull(mapView.getMapboxMap().getStyle()) {
+            _mapStyleSink.tryEmit(it)
+        }
+    }
+
+    private val onMapLongClickListener = OnMapLongClickListener { point ->
+        _mapLongClickSink.tryEmit(point)
+        false
+    }
+
+    private val onMapClickListener = OnMapClickListener { point ->
+        _mapClickSink.tryEmit(point)
+        false
+    }
+
+    private val onPositionChangedListener = OnIndicatorPositionChangedListener { point ->
+        jobControl.scope.launch {
+            _positionChangeSink.emit(point)
+        }
+    }
+
+    private val onMoveListener = object : OnMoveListener {
+        override fun onMove(detector: MoveGestureDetector): Boolean {
+            return false
+        }
+
+        override fun onMoveBegin(detector: MoveGestureDetector) { }
+
+        override fun onMoveEnd(detector: MoveGestureDetector) {
+            _mapMovementsSink.tryEmit(mapView.getMapboxMap().cameraState)
```

Comment:
Curious to find out why are we emitting the camera state when moving the map? `OnMoveListener` does this include zooming the Map as well?

Reply:
Moving the map changes the camera state. It makes sense the camera's state would be emitted.  While the recenter button might not care about the end state other components might.  Any component interested in map movements can use as much or as little of the camera state as it needs.

Zooming the map fires the OnMoveListener.

Reply:
Agreed that camera state changes on moving the map, but `MapEventDelegate` shouldn't be responsible to publish them in my opinion. I see it from a perspective where it's not the responsibility of the map to publish any changes to camera state, rather it belongs to `CameraCoordinator` or similar entity when we have one. Any UI component interested in knowing about changes in camera state should register with camera coordinator. On the contrary, when we move the map, I think it's even better to publish the offset by which the map has moved. If that's not available, just publishing `MOVE_STARTED` and `MOVE_ENDED` should be more than sufficient. This way we will have `MapEventDelegate` only responsible for communicating that map moved, whereas camera coordinator would be responsible for communicating that this is the new camera state because map was moved.

Reply:
This is an unnecessary adding of complexity.  A map move changes the map so the component should communicate the result of that move.  The component has been changed to express the MotionEvent which is useless information to the drop in components IMO.